File: week12/59-spiral-matrix-ii/generateMatrix.scala
object Solution {
  def generateMatrix(n: Int): Array[Array[Int]] = {
    val board = Array.ofDim[Int](n, n)
    val g = 1.to(n * n).iterator
    val range = 0 until n
    range.zip(range.reverse).foreach { case (up, down) =>
      up to down foreach (board(up)(_) = g.next())
      (up + 1) to down foreach (board(_)(down) = g.next())
      (down - 1).to(up, -1) foreach (board(down)(_) = g.next())
      (down - 1).until(up, -1) foreach (board(_)(up) = g.next())
    }
    board
  }
}

File: week12/59-spiral-matrix-ii/generateMatrix.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> board(n, vector<int>(n));
        int g = 1;
        for (int up = 0, down = n - 1; up < n; ++up, --down) {
            for (int i = up; i <= down; ++i) board[up][i] = g++;
            for (int i = up + 1; i <= down; ++i) board[i][down] = g++;
            for (int i = down - 1; i >= up; --i) board[down][i] = g++;
            for (int i = down - 1; i > up; --i) board[i][up] = g++;
        }
        return board;
    }
};
